Kernel/Bus/USB/UHCI/UHCIController.cpp

```diff
@@ -387,6 +389,37 @@ void UHCIController::dequeue_qh(QueueHead* transfer_queue)
     transfer_queue->prev_qh()->link_next_queue_head(transfer_queue->next_qh());
 }
 
+ErrorOr<QueueHead*> UHCIController::create_transfer_queue(Transfer& transfer)
+{
+    Pipe& pipe = transfer.pipe();
+
+    // Create a new descriptor chain
+    TransferDescriptor* last_data_descriptor;
+    TransferDescriptor* data_descriptor_chain;
+    auto buffer_address = Ptr32<u8>(transfer.buffer_physical().as_ptr());
+    TRY(create_chain(pipe, transfer.pipe().direction() == Pipe::Direction::In ? PacketID::IN : PacketID::OUT, buffer_address, pipe.max_packet_size(), transfer.transfer_data_size(), &data_descriptor_chain, &last_data_descriptor));
+
+    last_data_descriptor->terminate();
+
+    if constexpr (UHCI_VERBOSE_DEBUG) {
+        if (data_descriptor_chain) {
+            dbgln("Data TD");
+            data_descriptor_chain->print();
+        }
+    }
+
+    QueueHead* transfer_queue = allocate_queue_head();
+    if (!transfer_queue) {
+        free_descriptor_chain(data_descriptor_chain);
```

Comment:
This explicit free feels like we're not leveraging RAII enough in this subsystem. I hope there's a matching free somewhere else after the transfer is completed, or in the destructor of the transfer queue?

Reply:
Yeah this would be a good application of RAII, all of the cleanup is the responsibility of cancel_async_transfer currently (for async transfers).